system/security/plugins/jwtSecurity/README.md

```diff
@@ -0,0 +1,164 @@
+* If the token expires, the plugin will call the ```JWT refresh service``` to request a new token.  This follows
+<https://openid.net/specs/openid-connect-core-1_0.html#RefreshTokens> except that the ```client_secret``` and ```scope```
+fields in the request are omitted.
+
+### Implications of Deviations
+
+The most obvious outcome of this implementation is that a custom service/endpoint needs to be available.  Or rather two
+services:  One to handle the initial user login and one to handle token refreshes.  Neither service *precisely* handles
+requests and replies in an OIC-compatible way, but the tokens themselves *are* OIC-compatible, which is good.  That
+allows you to use third-party JWT libraries to construct and validate those tokens.
+
+### HPCC Systems Configuration Notes
+
+Several items must be defined in the platform's configuration.  Within configmgr, the ```jwtsecmgr``` Security Manager
+plugin must be added as a component and then modified according to your environment:
+
+* The URL or unique name of this HPCC Systems cluster, used as the ```client_id``` in token requests
+* Full URL to the ```JWT Login Endpoint``` (should be https, but not required)
+* Full URL to the ```JWT Refresh Endpoint``` (should be https, but not required)
+* Boolean indicating whether to accept self-signed certificates for those endpoints
+* Full pathname to the key file
+* Default workunit scope access level (either "Full" or "None")
+* Default file scope access level (either "Full" or "None")
+
+Only the first three items have no default values and must be supplied.
+
+Once the ```jwtseccmgr``` component is added, you have to tell other parts of the system to use the plugin.  For
+user authentication and permissions affecting features and workunit scopes, you need to add the plugin to the ```esp```
+component.  Instructions for doing so can be found in the HPCC Systems Administrator's Guide manual (though the
+manual uses the htpasswd plugin as an example, the process is the same).
+
+If you intend to implement file scope permissions then you will also need provide Dali information about the JWT
+plugin.  In configmgr, within the ```Dali Server``` component, select the LDAP tab.  Change the ```authMethod``` entry
+to ```secmgrPlugin``` and enter "jwtsecmgr" as the ```authPluginType```.  Make sure ```checkScopeScans``` is set to
+true.
+
+### HPCC Systems Authorization and JWT Claims
+
+This plugin supports all authorizations documented in the HPCC Systems® Administrator's Guide with the exception of
+"View Permissions".  Loosely speaking, the permissions are divided into three groups:  Feature, Workunit Scope, and File
+Scope.
+
+Feature permissions are supported exactly as documented.  A specific permission would exist as a JWT claim, by name,
+with the associated value being the name of the permission.  For example, to grant read-only access to ECL Watch, use
```

Comment:
In the future if we support other sources of JWTs and have less control over say the naming of claims or structure of the token we should implement mapping that says "for permission xyz the path to look at in the token is a.b.c and the value required is x".  That is the purpose of our mapping files now, but if we support multiple sources of jwts then we may need one per source.

Reply:
My point being that when we can dictate the JWT format / claims as we can now, that's great.. we may not always be able to in the future.

Reply:
I think this is something we should revisit only if it comes up.  What you say makes sense, but I'm not convinced the need for a map like that will arise in the real world.  I'd like to see a concrete example before embarking on a mapping implementation.

Reply:
Yes, we can revisit if and when needed.